lib/SILOptimizer/Mandatory/DefiniteInitialization.cpp

```diff
@@ -783,6 +783,162 @@ void LifetimeChecker::diagnoseInitError(const DIMemoryUse &Use,
+          break;
+        }
+      }
+    }
+  }
+
+  if (!B.hasValidInsertionPoint())
+    llvm::report_fatal_error("did not find InitializeDefaultActor built-in!");
+
+  // get the already-borrowed `self` value from the built-in's operand.
+  assert(Inst->getNumOperands() == 1
+          && "unexpected number of operands for InitializeDefaultActor");
+  auto &firstOpr = Inst->getAllOperands().front();
+
+  // insert the hop_to_executor
+  B.createHopToExecutor(ctor, firstOpr.get(), false);
+}
+
+void LifetimeChecker::injectActorHops() {
+  auto ctor = TheMemory.isActorInitSelf();
+
+  // Must be `self` within an actor's designated initializer.
+  if (!ctor)
+    return;
+
+  // If the initializer has restricted use of `self`, then no hops are needed.
+  if (isRestrictedActorInitSelf())
+    return;
+
+  // Even if there are no stored properties to initialize, we still need a hop.
+  if (TheMemory.getNumElements() == 0)
+    return injectHopAfterBuiltinInit(ctor,F.getEntryBlock());
+
+  // A mapping from initializing blocks to their in-availability set.
+  using Map = llvm::SmallDenseMap<SILBasicBlock*, AvailabilitySet, 8>;
+
+  /////
+  // Step 1: Find initializing blocks, which are blocks that contain a store
+  // to TheMemory that fully-initializes it, and build the Map.
+
+  Map initializingBlocks;
+  {
+    llvm::SmallDenseSet<SILBasicBlock *> seenBlocks; // processed blocks
+
+    // We determine whether a block is "initializing" by inspecting the "in" and
+    // "out" availability sets of the block. If the block goes from No / Partial
+    // "in" to Yes "out", then some instruction in the block caused TheMemory to
+    // become fully-initialized, so we record that block and its instructions
+    // that might be the ones that fully-initialized TheMemory.
+    for (auto const &entry : NonLoadUses) {
+      auto *inst = entry.getFirst();
+      auto *block = inst->getParent();
+
+      // TODO: we could probably skip use kinds that do not initialize memory.
+
+      // If we have already seen this block...
+      if (!seenBlocks.insert(block).second)
+        continue; // skip it!
+
+      // Determine if this block is initializing, that is:
+      //
+      //     merge(OutAvailability(block.pred1, block.pred2, ...)) != Yes
+      //               AND
+      //     OutAvailability(block)) = Yes
+
+      auto outSet = getBlockInfo(block).OutAvailability;
+      if (!outSet.isAllYes())
+        continue; // then this block never sees TheMemory fully-initialized.
+
+      AvailabilitySet inSet(outSet.size());
+      for (const auto &pred : block->getPredecessorBlocks())
+        inSet.mergeIn(getBlockInfo(pred).OutAvailability);
+
+      if (inSet.isAllYes())
+        continue; // then this block always receives TheMemory fully-initialized.
+
+      initializingBlocks.insert({block, inSet});
+    } // end loop
+
+#ifndef NDEBUG
+    // dump the data-structure we just built
+    for (auto &entry : initializingBlocks) {
```

Comment:
Please hide this behind LLVM_DEBUG. People compile with asserts all the time and they shouldn't have to see this.